src/commons/navigationBar/subcomponents/SicpNavigationBar.tsx

```diff
@@ -222,212 +111,541 @@ const SicpNavigationBar: React.FC = () => {
       )
       .map(word => pre + ' ' + word);
   }
-
-  const handleSearchButton = () => {
-    handleOpenSearch();
-    setDisplayedQuery(searchQuery);
-    setQueryResult(voidSearch(searchQuery));
+  */
+  // this section responsible for the search
+  type TrieNode = {
+    children: Record<string, TrieNode>;
+    value: any[];
+    key: string;
   };
-  const handleAutoSearch = (str: string) => {
-    handleOpenSearch();
-    setDisplayedQuery(str);
-    setQueryResult(voidSearch(str));
-    setSearchQuery(str);
-  };
-  const handleAutoIndexSearch = (str: string) => {
-    handleIndexSearchButton(str);
+
+  type SearchData = {
+    indexTrie: TrieNode;
+    textTrie: TrieNode;
+    idToContentMap: Record<string, string>;
   };
 
-  const handleIndexSearchButton = (str: string) => {
-    handleOpenSearch();
-    setDisplayedQuery(str);
-    const SearchUrl = '..';
-    const tem = [];
-    const ans = queryTrie(indexTrie, str.toLowerCase());
-    if (ans == null) {
-      tem.push({ title: 'no result found', url: '' });
+  const fetchSearchData = () => {
+    const xhr = new XMLHttpRequest();
+    const url = Constants.sicpBackendUrl + 'json/rewritedSearchData.json';
+    xhr.open('GET', url, false); //sync download
+    xhr.send();
+    if (xhr.status !== 200) {
+      alert('Unable to get rewrited search data. Error code = ' + xhr.status + ' url is ' + url);
+      throw new Error('Unable to get search data. Error code = ' + xhr.status + ' url is ' + url);
     } else {
-      const pure = ans['pureIndex'];
-      for (let i = 0; i < pure.length; i++) {
-        // pure[i][0] is sth like /sicpjs/3.3.3; slice out the /sicpjs/
-        tem.push({
-          title: pure[i][0].slice(8) + ': ' + ans['value'],
-          url: SearchUrl + pure[i][0] + pure[i][1]
-        });
+      const searchData: SearchData = JSON.parse(xhr.responseText);
+      return searchData;
+    }
+  };
+
+  function search(keyStr: String, trie: TrieNode) {
+    const keys = [...keyStr];
+    let node = trie;
+    for (let i = 0; i < keys.length; i++) {
+      if (node === undefined || node.children === undefined) {
```

Comment:
Likewise over here, the way `TrieNode` is defined suggests that `node.children` will never be undefined, so can I double-check on what the proper structure of this object is supposed to be?

Reply:
It should not be undefined. I add this clause to be defensive